Many improvements to docker network and container states

terminalmage opened the pull request:
Much Improved Support for Docker Networking
===========================================

The `docker_network.present` state has undergone a full rewrite, which includes the following improvements:

Full API Support for Network Management
---------------------------------------

The improvements made to input handling in the `docker_container.running` state for 2017.7.0 have now been expanded to `docker_network.present`.  This brings with it full support for all tunable configuration arguments.

Custom Subnets
--------------

Custom subnets can now be configured. Both IPv4 and mixed IPv4/IPv6 networks are supported.

Network Configuration in `docker_container.running` States
-------------------------------------------------------------------

It is now possible to configure static IPv4/IPv6 addresses, as well as links and labels.

Improved Handling of Images from Custom Registries
==================================================

Rather than attempting to parse the tag from the passed image name, Salt will now resolve that tag down to an image ID and use that ID instead.

Due to this change, there are some backward-incompatible changes to image management. See below for a full list of these changes.

Backward-incompatible Changes to Docker Image Management
--------------------------------------------------------

Passing image names to the following functions must now be done using separate
`repository` and `tag` arguments:

- `docker.build`
- `docker.commit`
- `docker.import`
- `docker.load`
- `docker.tag`
- `docker.sls_build`

Additionally, the `tag` argument must now be explicitly passed to the `docker_image.present` state, unless the image is being pulled from a docker registry.

"docker run" Support Added
==================================================

A `docker.run_container` function has been added to mimic `docker run`, running the container and returning the logs from it. In addition, a `docker_container.run` state has been added, which includes many of the same features as the `cmd.run` state (including `onlyif`, `unless`, and `creates` arguments to limit whether or not the container is actually run).

# Tests written?

You betcha! Dozens of new tests, and rewrites of most of the rest due to abstraction of the input translation code.

# Commits signed with GPG?

Yes

terminalmage committed:
Many improvements to docker network and container states

Much Improved Support for Docker Networking
===========================================

The `docker_network.present` state has undergone a full rewrite, which
includes the following improvements:

Full API Support for Network Management
---------------------------------------

The improvements made to input handling in the
`docker_container.running` state for 2017.7.0 have now been expanded to
docker_network.present`.  This brings with it full support for all
tunable configuration arguments.

Custom Subnets
--------------

Custom subnets can now be configured. Both IPv4 and mixed IPv4/IPv6
networks are supported.

Network Configuration in :py:func:`docker_container.running` States
-------------------------------------------------------------------

It is now possible to configure static IPv4/IPv6 addresses, as well as
links and labels.

Improved Handling of Images from Custom Registries
==================================================

Rather than attempting to parse the tag from the passed image name, Salt
will now resolve that tag down to an image ID and use that ID instead.

Due to this change, there are some backward-incompatible changes to
image management. See below for a full list of these changes.

Backward-incompatible Changes to Docker Image Management
--------------------------------------------------------

Passing image names to the following functions must now be done using separate
`repository` and `tag` arguments:

- `docker.build`
- `docker.commit`
- `docker.import`
- `docker.load`
- `docker.tag`
- `docker.sls_build`

Additionally, the `tag` argument must now be explicitly passed to the
`docker_image.present` state, unless the image is being pulled from a
docker registry.

terminalmage commented:
Note: the integration tests for docker containers and networks are being skipped on Ubuntu 14 because salt-jenkins is not installing docker, docker-py, and busybox like it does for Cent7. Is this something we want to add to salt-jenkins for Ubuntu?

terminalmage committed:
Touch-ups to docstring

terminalmage committed:
Add docker_container.run state and docker.run_container func

rallytime commented:
@terminalmage I don't know of any reason off-hand of why we would be limiting those tests to CentOS 7. If you wanted to add that to salt-jenkins, I think that would be just fine.

terminalmage committed:
Update busybox image script to include /etc/{passwd,group,shadow}

These are necessary to make sure users work

terminalmage committed:
Add integration tests for docker_container.run state

terminalmage commented:
https://github.com/saltstack/salt-jenkins/pull/573 adds docker to salt-jenkins for Ubuntu 16.04. I tested this in a vagrant box and all tests passed.

terminalmage committed:
Update release notes for new "docker run" functionality

rallytime commented:
Go Go Jenkins!

terminalmage commented:
This has been rebased to include fixes for the unrelated test failures that were merged in the last 24 hours.

terminalmage commented:
There were some lingering test failures which I believe I have resolved. Tests running again now.

terminalmage committed:
Add debug logging to display the API version

This will help troubleshoot when an API version requirement is not met.

rallytime commented:
@terminalmage There are 2 docker tests failing on the cent7 build:

- integration.states.test_docker_container.DockerContainerTestCase.test_run_bg
- integration.states.test_docker_network.DockerNetworkTestCase.test_present_attachable

Can you take a look?

terminalmage committed:
Remove use of auto_remove in docker_container integration tests

CentOS 7 still ships a version of Docker with API version 1.24 which
does not support this argument. This commit uses `docker.rm` to remove
the container manually after the state is run, in cases where we were
using auto_remove=True.

terminalmage committed:
Handle the fact that attachable is unmanageable on Cent7

terminalmage committed:
Use __utils__ instead of salt.utils for function calls

terminalmage committed:
More conversion to __utils__

Also, actually using __utils__ here revealed several import errors in
utils modules, so these needed to be dealt with.

cachedout commented on doc/topics/releases/oxygen.rst:
```diff
@@ -4,6 +4,109 @@
+A long-requested feature has finally been added! It is now possible to
+configure static IPv4/IPv6 addresses, as well as links and labels. See
+:ref:`here <salt-states-docker-container-network-management>` for more
+informaion.
```
`information`

terminalmage replied:
/me should have used `:set spell`. 😄 

Thanks!

cachedout commented on doc/topics/releases/oxygen.rst:
```diff
@@ -4,6 +4,109 @@
+Full API Support for :py:func:`docker.logs <salt.modules.dockermod.logs>`
+=========================================================================
+
+This function now supports all of the functions that it's Docker API
```
`its`

terminalmage committed:
Grammar/spelling fixes in release notes

terminalmage committed:
remove blank lines to satisfy linter

rallytime reviewed (approved):
There is a lot here, but this is really well done @terminalmage. Nice work.
